### DISCUSSION TITLE
Support for Dynu v2 API

### DISCUSSION THREAD

**Discussion by cross** [troglobit/inadyn#491] [Ideas]
I _think_ dynu.com no longer supports a simple access method.  Their REST API is documented at https://www.dynu.com/Support/API. I am planing to copy cloudflare/porkbun plugins into a new dynu plugin.

Starting this discussion thread so I can write questions and get opinions from core maintainers.  Thanks.

**Comment by cross** [comment]
First question is, dynu has a "give me all of my domains" API, and a "get me the domain that hostname FOO is in".  Using the latter with the specified hostname would be the easiest way to get the domain ID, needed for most calls.  This would fail if the hostname didn't already exist.  So, it seems less "correct".    More correct to get all of the domains and match for the given domain name (given as with cloud flare zone in the username field), but that's a bunch more code (and doesn't match existing code, so more work)

Opinions welcome, and any pointers to code that I haven't noticed yet that will search JSON responses for matching objects, rather than a single matching string.

**Reply by cross** [reply]
Alright.  I think I've talked myself down from the "I must parse through the domains array."  It occurred to me that the actual domain name itself is almost certainly (maybe even guaranteed?) to be a "host" record inside of the domain.  So I will query for that to get the domain id.  Leaving this here just for posterity and possible comment.

**Comment by acolomb** [comment]
Are you sure that we actually need to do anything more complicated than what the plugin does now?  I'd need to check if all is still working, but at least in my setups, I haven't noticed any changes / failures lately.  The "old" API still works perfectly, which is much simpler and basically modeled after the original DynDNS API: https://www.dynu.com/en-US/DynamicDNS/IP-Update-Protocol

Where did you get the impression that it no longer works?

**Comment by cross** [comment]
Yeah, I've been thinking about that too.  It no longer worked for me.  It's been a few years, and I see a bunch of text on their site that suggests something was now unsupported.  But, I have been talking to their support folks about the "old" API, which is still documented, and they have finally gotten me enough info to make it work.  At least by hand with curl.  I was going to go back to seeing if the dyndns plugin would work.
One thing I've found talking to their support is that creating _new_ hosts is not possible with the old API.  And whether or not updating IPv6 addresses works is unclear, it sounds from what they say that it [probably?] works but doesn't report that it works.

Anyway.  Thanks aclomob for the feedback, do let me know if you find it is still working.  And are you using "hostname" or "alias" parameters?  (The old API only accepts "alias" parameters for hostnames, it's weird, but I think I know the details of it).  I am setting half a dozen hostnames anytime I update my address, so figuring out which syntax to use is something I haven't figured out in inadyn yet.  I just yesterday got so I could make the updates with curl by hand and see any success.

**Reply by cross** [reply]
I just took a closer look at `plugins/dyndns.c` and `plugins/common.c`.  That will not work for me.  I think if someone had a domain for which only their main registration needed an address, it would work.  But, talking to their support and testing with curl, I can confirm that for hosts within my domain, "hostname=alias.domain.com" does _not_ work.  The old API requires "hostname=domain.name&alias=alias" which the dyndns plugin doesn't do.

I'm guessing this is why it seemed to stop working for me at some point.  Whether this justifies a new plugin is arguable, I agree.

**Reply by acolomb** [reply]
For reference, here is my config snippet for an alias name at dynu.com.  Note that it uses some special syntax because I need an IPv6 AAAA record specifically and that is not supported for a third-level zone directly (you cannot disable its A record).  The host name and alias need to be adjusted in two spots, of course.  But in principle it works fine, also for A record aliases if you need that.

```
custom dynu.com:ipv6-myhost {
        hostname        = ipv6.myhost.dynu.net
        username        = itsme
        password        = XXX
        ddns-server     = api.dynu.com
        ddns-path       = "/nic/update?hostname=myhost.dynu.net&alias=ipv6&myipv6="
        append-myip     = true
        ddns-response   = { "good", "nochg" }
}
```

**Reply by cross** [reply]
If I read that URL right, that will set an alias within the hosted domain of "myhost.dynu.net".  So, set an AAAA record for `ipv6.myhost.dynu.com`.  That's at least how I've seen things.  Is the domain you're hosting listed in the control panel as "myhost.dynu.com" (for this example)?
In my testing, if I set "hostname=domain.com", where domain.com is the single DDNS domain I have hosted with this, I can then set the A (or AAAA) record of a host inside of it with "alias=myhost".  But, if I set "hostname=myhost.domain.com&alias=myhost", then it will _report_ success to the API call, but not make the change.
I will contact your via the email listed in your profile and we can talk more detail.  Thank you for your help.

**Reply by cross** [reply]
Also, the above is _not_ using the dyndns provider's support of dynu, correct?  It's purely using the custom provider codepaths.
Let me know if I'm wrong, but if I'm correct, this supports my suggestion that the non-working dynu.com 
"support" be taken out of plugins/dyndns.

**Reply by acolomb** [reply]
The plugin is working fine AFAIK.  It just doesn't support the add-on "alias" feature of Dynu.  It would be a welcome addition to support that without resorting to a custom provider.  Feel free to dig into it and make a PR.  But that extra feature being missing doesn't warrant ripping out code that works for the more common use cases.

**Reply by cross** [reply]
That's my question though.  Is your config listed above using the dyndns plugin?  From what the support folks at dynu said, `hostname` must be the registered DDNS domain, and records within that require the `alias=` syntax.  That latter, as you note, is not supported by the dyndns plugin.  So unless someone is updating their primary domain's A/AAAA record I think dyndns plugin won't work for dynu.com.

As to adding support for `alias` parameter that dynu.com uses, I'd be happy to do that, but I have no way to test (and likely not enough documentation) to verify that it would work for any or all of the other providers that are supported by the dyndns plugin.  So I don't know that that would be practical.   If I'm being overly pessimistic and you think it is a frequently supported feature of this API, let me know.

**Reply by acolomb** [reply]
Quoting some stuff that I answered to @cross directly...

> How my solution works is quite simple: Give Inadyn the actual DNS name that you want it to check the IP for.  That is the alias name, concatenated with the domain name, as shown in the Aliases table.
>
> But the update URL must contain the two pieces of information separately, which Inadyn cannot provide.  Therefore I override the ddns-path and simply put the two values in there directly, without automatic substitution.  Only the IP address is inserted at the needed place.  Which can be done using the %i format specifier within the path, or by ending the path with an empty IP parameter and setting the "append-myip = true" option in addition.  Note that you don't need the username field in the URL, it is passed together with the password using HTTP Basic authentication.
>
> Adding the alias parameter to the dynu.com Inadyn plugin properly would be a nice addition, with an associated custom option.  But I haven't found the spare time to look into that, as the custom provider workaround is enough for me.  Your contribution to integrate it would be very welcome, without switching to a whole different API though.  I'd be happy to test any patches. 

As for the plugin being used, I haven't dug into the code, but I thought the dynu plugin was somehow derived from the dyndns plugin, just overriding some methods / default values?  I think the feature should be added in just that way, as an extension within the dynu.com plugin, which inherits all other logic from the dyndns base.  There should be some examples of such additions in the code already, IIRC.  Note that what Inadyn's code calls an "alias" internally has nothing to do with the Dynu concept, though.

**Reply by cross** [reply]
If you say "provider default@dynu.com" it uses plugins/dyndns.c.  It just uses a structure that defines a different checkin and api server name.  But the syntax of the parameters to the API server and path are the same for all providers supported by plugins/dyndns.  It's all one plugin, just with different default values, but no ability to change code-path.

**Reply by acolomb** [reply]
Well then you probably need a new plugin, which re-uses most of the common request / response functions though.  I'm not sure how custom config options can be added for a new plugin.  Maybe @troglobit can give a quick pointer here?  What the cloudflare plugin does is store some additional data in the `info->data = (struct cfdata *) ...` structure.  But from a quick glance, it seems the data there comes from another HTTP request within the `setUp()` function, not from additional config options.

**Reply by acolomb** [reply]
I don't see how it could be done without a new config option.  Of course, we could just make a new provider plugin "dynu.com-alias" that simply splits each given hostname at the first dot, then puts the two parts into the alias and hostname URL fields, respectively.   But that won't work for aliases containing a dot themselves.  You somehow need to communicate which part is the domain and where the alias ends.

I guess repeating the fqdn for each alias in the `hostname` parameter would be the simplest approach, because all the DNS lookup and caching logic won't need to be changed.  Then add another `common_zone` or whatever provider option that is stripped from the end of each fqdn to separate alias from hostname.  If a hostname in that provider config section doesn't have that common suffix, it is simply not split, but treated as before.  That would make the fallback behavior identical to the current functionality.

**Reply by cross** [reply]
Side note, while looking around for docs of this API by other providers, I found the following discussion at ddns-updater:

https://github.com/qdm12/ddns-updater/discussions/385

It looks like someone had a problem some time ago that needed _not_ to use the alias= syntax, and a change was put in (https://github.com/qdm12/ddns-updater/commit/83a9a2d999392dfcaee1d0b0e3dab538d52c337c) that recently caused someone else the same problem I'm now having.

I'm not sure this helps us here, but it is more discussion of the complexities with dynu, and qdm12 asks good questions about what syntax should be used for what situations.

**Reply by cross** [reply]
> I don't see how it could be done without a new config option. Of course, we could just make a new provider plugin "dynu.com-alias" that simply splits each given hostname at the first dot, then puts the two parts into the alias and hostname URL fields, respectively. [...]

At this point, we're back to talking about a new plugin, which is where I started.  :-)  I would argue _against_ duplicating the Dyn protocol plugin into something else specific for the Dynu variant that has an "alias" parameter.  Unless there is more than one provider that does this with their equivalent to `/nic/update`.  If a new plugin is warranted, and dynu.com only, I think using the REST API they suggest is the better plan.  But not a question to answer now.  For now let's stay focused on "does the existing dyndns plugin work for dynu.com, and in which use cases?"

**Reply by acolomb** [reply]
It works. Even the generic (custom provider) plugin works. To make use of their alias feature, the custom provider and some tricks regarding the settings are needed.

I don't see how implementing a whole new API should be better than just some string splitting and putting that into a new request URL template. And that's not just judging the time to get a working implementation, but rather the effort needed to maintain it in the future, besides the API that's used by probably > 90 % of the available providers in Inadyn. If adding to the dyndns plugin is the most effective way, then let's do that and have all other providers simply not use the extra feature.

**Reply by cross** [reply]
I'm still not sure you can say that the dyndns plugin works for dynu.com.  There are at least use cases where it cannot work.  I was just saying that using the REST API would be preferable to making a whole separate copy of the dyndns plugin just for Dynu.  I agree with you that if we can make the dyndns plugin support dynu more fully, that is a better option.

**Reply by acolomb** [reply]
Do you realize that it's not really about a "whole separate copy of the dyndns plugin", but most of the functionality can be reused?  Have a look at https://github.com/troglobit/inadyn/blob/master/plugins/all-inkl.c for example, it basically just adapts to some different parameter names in the URL string template.  Everything else is still handled by the common code from the dyndns plugin.

**Reply by cross** [reply]
Following up on this, I have learned more about the use cases and API parameters from https://github.com/qdm12/ddns-updater/discussions/385.   So, as it is, a custom config is needed when you're updating dynu for anything inside of the given domain they're providing.  It is not possible, with the existing dyndns implementation for dynu, to set underlying host address records.  I still mostly feel that this means the dynu support in dyndns is _not_ working.  It fails for a common use case.  It works for another common use case, but not mine.
I still think that using a more complete recommended API is the better solution, but accept that you do not, and sortof understand your reasons. (reasons notwithstanding, I agree your decisions are the more important ones. :-) )
I will see if I can make a simplistic implementation around dyndns/generic.  I do not understand from a quick look at `all-inkl` host to do that, but I will ponder it more.

**Comment by cross** [comment]
So for @troglobit (and anyone else who makes such decisions), it seems to me that the existing dyndns doesn't work well enough for dynu.  Even if adding records isn't needed, I presume wanting to set more than one hostname, and hostnames _inside_ the domain level, is something you/we want to support.
If I removed the dynu lines from dyndns code, and added a new REST API plugin, would that be a PR you'd accept?
I'd also refactor the common code out that is shared between cloudflare and porkbun, which three plugins would now use.

**Reply by acolomb** [reply]
Please don't remove the dyndns based support for Dynu. It has the big advantage that the API is so widespread that it doesn't really evolve anymore. Just some proprietary extensions here and there, like the alias function.

Any REST API is in my experience more complex and often a moving target, where new versions come and old ones are deprecated at some point. I'd be happy to see native support for it in inadyn, but will you also be there to keep it updated for the foreseeable future?

I actually am using Dynu successfully with inadyn, even with aliases. But I put a custom provider config section in the file instead of specifying dynu.com as provider. That works fine with the alias extension, you just need to put the alias and zone in the URL directly, while keeping the hostname field with both concatenated (so the DNS lookup works properly).

**Reply by troglobit** [reply]
> Please don't remove the dyndns based support for Dynu. It has the big advantage that the API is so widespread that it doesn't really evolve anymore. Just some proprietary extensions here and there, like the alias function.
> 
> Any REST API is in my experience more complex and often a moving target, where new versions come and old ones are deprecated at some point. I'd be happy to see native support for it in inadyn, but will you also be there to keep it updated for the foreseeable future?

+1 on all of the above! Thank you @acolomb 🙏

**Reply by cross** [reply]
Following up here, I'm trying to make a config that works for my with dynu.com.  While doing this, I notice that I get successful response to the `/nic/update` URI regardless of whether it's really successful.  Is this just me?
For example:
```
cross@router: curl -X GET --basic -u bob@goo.com  "https://api.dynu.com/nic/update?hostname=foo.com&alias=bobcat&myip=192.118.45.123"
Enter host password for user 'bob@goo.com':
good 192.118.45.123%
cross@router: 
```
No valid account, no valid password, reports success.  I've no idea why this is true, since I know I was getting "badauth" responses with this sort of manual request weeks ago.  But, this makes it hard for me to test whether a config works, since everything reports success.

**Reply by troglobit** [reply]
> Following up here, I'm trying to make a config that works for my with dynu.com.  While doing this, I notice that I get successful response to the `/nic/update` URI regardless of whether it's really successful.  Is this just me?
> 
> For example:
> 
> ```
> 
> cross@router: curl -X GET --basic -u bob@goo.com  "https://api.dynu.com/nic/update?hostname=foo.com&alias=bobcat&myip=192.118.45.123"
> 
> Enter host password for user 'bob@goo.com':
> 
> good 192.118.45.123%
> 
> cross@router: 
> 
> ```
> 
> No valid account, no valid password, reports success.  I've no idea why this is true, since I know I was getting "badauth" responses with this sort of manual request weeks ago.  But, this makes it hard for me to test whether a config works, since everything reports success.

This sounds more like a question for Dynu rather than us, right?

**Reply by cross** [reply]
Yeah, I've asked them.  Sorry.  Lots of testing going on, hard to figure out what does and doesn't work.

**Comment by cross** [comment]
Thank you for your feedback.  I'll abandon my REST API code then.  Thank you @acolomb for your help in getting the custom config working for my needs.
